pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package abcupdater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/abc-updater/pkg/abcupdater/localstore"
+	"github.com/abcxyz/pkg/logging"
+)
+
+const (
+	installIDFileName = "id.json"
+)
+
+var regExUUID = regexp.MustCompile("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
+
+type metricsConfig struct {
+	ServerURL string `env:"ABC_UPDATER_METRICS_URL,default=https://abc-updater-metrics.tycho.joonix.net"`
+}
+
+// MetricsInfo are the parameters for sending metrics.
+type MetricsInfo struct {
+	// The ID of the application to check.
+	AppID string
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string
+
+	// An optional Lookuper to load envconfig structs. Will default to os environment variables.
+	Lookuper envconfig.Lookuper
+
+	Metrics map[string]int
+
+	// Optional override for install id file location. Mostly intended for testing.
+	// If empty uses default location.
+	InstallIDFileOverride string
+}
+
+// InstallIDData defines the json file that defines installation id.
+type InstallIDData struct {
+	// Time ID was created, in UTC epoch seconds.
+	IDCreatedTimestamp int64 `json:"idCreatedTimestamp"`
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+type SendMetricRequest struct {
+	// The ID of the application to check.
+	AppID string `json:"appId"`
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string `json:"version"`
+
+	Metrics map[string]int `json:"metrics"`
+
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+// Stricter than uuid.Parse() which isn't meant for validating strings according
+// to documentation.
+func validInstallID(id string) bool {
+	return regExUUID.MatchString(id)
+}
+
+// SendMetricsSync sends information about application usage. Users can opt out
+// by setting an env variable defined in opt_out.go.
+// Accepts a context for cancellation.
+func SendMetricsSync(ctx context.Context, info *MetricsInfo) error {
+	lookuper := info.Lookuper
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
+	}
+
+	optOutSettings, err := loadOptOutSettings(ctx, lookuper, info.AppID)
+	if err != nil {
+		return fmt.Errorf("failed to load opt out settings: %w", err)
+	}
+
+	if optOutSettings.NoMetrics {
+		return nil
+	}
+
+	generateNewID := true
+	storedID, err := loadInstallID(info)
+	if err == nil && storedID != nil {
+		oneDayAgo := time.Now().Add(-24 * time.Hour)
+		// Defensively check for case ID Created is in future.
```

Comment:
Sorry if this is already made clear elsewhere, but why is it bad to have a future-dated ID?

Reply:
The rolling install ID is meant as a privacy measure. In theory if a misconfigured client generated an ID with a date 1000 years into the future, a new install id would never be generated again, violating the privacy measure.

The installID is intended to avoid issues where one badly behaving or busy client sends a large number of requests and we are left with no way to separate that from actual metrics. Its not meant for tracking how a particular user uses an app.

Reply:
This all seems overly complex? I'm not sure I understand why all this complexity is required.

Reply:
This has been simplified to a static install id